### DISCUSSION TITLE
Find subtitles in folders?

### DISCUSSION THREAD

**Discussion by Freethefire** [KBlixt/subcleaner#7] [Ideas]
Hey, thanks for a excellent script!  
I was wondering if it's possible to set this up so that it searches for new subtitles in folder and subfolders? I realize that I'd probably set up a log file for already processed files. But maybe that's out of the scope for this solution?
Thanks!

**Comment by KBlixt** [comment]
I'm a bit unsure what you mean?

Do you mean to add functionality that would skip subtitles that have been processed already when it runs through a library? 

If that's the request then what is needed is to add some sort of persistence to the script that alter its behaviour depending on previous runs. I'm currently considering adding persistence through a database. If I end up doing it it will enable a lot of nice features to the script. skipping processed subtitles would be one of those features. 

It's not super trivial though. How do I detect that a subtitle have been processed already? I can't just look at the file name since you could have downloaded a new subtitle since I looked at it last time. I would basically have to run a hash of the file and compare it to the hash the file had when I looked at it last time. 

I'll probably end up adding a database to the script in order to further enhance usability. But since I want to be careful about not breaking      how it currently handles I can't say when this will happen.

I'm not sure how you get your subtitles but there might be ways around the issue you are trying to solve though other means. i.e you could use `list` to run the script on files matching certain criterias. 

I'm glad you like the script!

**Reply by Freethefire** [reply]
I'm sorry if my request was unclear!
The thought was that it would be possible to run the script on a media drive/partition and only pick up new subtitles that haven't been processed before. That is, not point the script at a specific file/subtitle, but make a cronjob (for example) that runs every day and cleans the subtitles that has been added since the day before. What came to mind was the Filebot log files, listing files already renamed, I don't know if you're familiar with how it works.
Since I cannot code myself, I'm really unsure how much work that would take to accomplice. It was just a thought, but I think a lot of people would find it helpful :)

Thanks again and thanks for replying!

**Reply by KBlixt** [reply]
Like I said in my reply, the functionality you suggest is not super easy to implement but my goal is to add it eventually. 

but you can accomplish what you want through the linux `find` program. 

i.e:

`find /media/library -type f -name "*.srt" -mtime -1 -exec python3 /opt/subcleaner/subcleaner.py {} /;`

would find all files ending with srt under the directory /media/library that have changed within one day and then run the subcleaner script on all files it find. 

Don't use this as a cron script though since it will cascade every subtitle onwards. but you could filter files on other properties like owner, group and permissions and add the marker to the files after the script have been ran.

How do you add your subtitles? do you do it manually?

**Reply by Freethefire** [reply]
Oh, great thanks!

Well, I do use Bazarr (and saw your tips about that), but not exclusively. I also Jellyfin and sometimes add manually.
Edit: Oh and autosubs via Kodi as well.

**Reply by KBlixt** [reply]
aright, I'll let you know when I get around to add this :)

**Reply by Freethefire** [reply]
> `find /media/library -type f -name "*.srt" -mtime -1 -exec python3 /opt/subcleaner/subcleaner.py {} /;`

~~I realize that you're not a support for Linux commands, but I'm getting "find: missing argument to `-exec'" when running the above command, pointing it at a folder in my media library. If you feel like it I would love to know why it's not working :) Thanks.~~

It seems `/;` needed to be `\;` at the end, or `+`. So nevermind :)

**Reply by KBlixt** [reply]
Yeah, I mistyped, good you managed to solve it.

Again, remember that this script will filter on files changed in the past 24h and then change them. So every subtitle changed in this run will have been changed within 24h of the next time you run the script if you run it every 24h

**Comment by televisi** [comment]
Just wanted to say THANK YOU for such an awesome godsend script!
Just an idea, perhaps a good idea to put the information to run this script against existing srt file on the main README page, as I found this discussion is useful too.

The command I used: 
`find /movies -type f -name "*.srt" -exec python3 /config/script/subcleaner/subcleaner.py --dry-run {} \;`

**Reply by KBlixt** [reply]
Good Point, I'll make a small section for it tomorrow. 

the script already have a recursive mode though, the -r flag allow you to find all srt files in a directory recursively. so while the find method is fine, it's easier to use the recursive mode directly. But as you correctly point out, this info isn't in the README. it's just in the --help message.
 
And once again, I'm glad you enjoy it :)
